### PR TITLE
Shutdown implementation

### PR DESCRIPTION
**Pull request by ghostdogpr**
This PR implements #25 

Note that my test names conflicts with the other PRs, I'll update them if the other PRs are merged first. Some other adjustments will also be needed after #17 is merged (offerAll also needs to terminate).

### CODE REVIEW

File: src/main/scala/scalaz/ioqueue/Queue.scala [hunk 1]
```diff
@@ -99,13 +104,14 @@ class Queue[A] private (capacity: Int, ref: Ref[State[A]]) {
...
-    ref.modify[List[A]] {
+    IO.flatten(ref.modify[IO[Nothing, List[A]]] {
       case Surplus(values, putters) =>
         val (q1, q2) = values.splitAt(max)
 
-        (q1.toList, Surplus(q2, putters))
-      case state @ Deficit(_) => (Nil, state)
-    }
+        (IO.point(q1.toList), Surplus(q2, putters))
+      case state @ Deficit(_)       => (IO.point(Nil), state)
```
Comment (wi101) [inline src/main/scala/scalaz/ioqueue/Queue.scala]: You can use `IO.now(Nil)` instead

File: src/main/scala/scalaz/ioqueue/Queue.scala [hunk 2]
```diff
@@ -137,6 +143,48 @@ class Queue[A] private (capacity: Int, ref: Ref[State[A]]) {
...
+          val forked = IO
+            .forkAll[Nothing, Boolean](
+              takers.toList.map(p => interruptPromise(p, l))
+            )
+            .flatMap(_.join)
+          (forked, Shutdown(l))
+        case state @ Shutdown(_) => (IO.unit, state)
+        case _                   => (IO.unit, Shutdown(l))
+      }
+      .flatMap(identity)
```
Comment (wi101) [inline src/main/scala/scalaz/ioqueue/Queue.scala]: why did you need to `.flatMap(identity)` and not directly use `.void` ?

**Comment by wi101** [comment]
@ghostdogpr Good job! :) 🙏

File: src/main/scala/scalaz/ioqueue/Queue.scala (hunk 1 above)
Comment (ghostdogpr) [reply src/main/scala/scalaz/ioqueue/Queue.scala]: Right, I'll change it.

File: src/main/scala/scalaz/ioqueue/Queue.scala (hunk 2 above)
Comment (ghostdogpr) [reply src/main/scala/scalaz/ioqueue/Queue.scala]: Here ref.modify returns an IO[Nothing, IO[Nothing, Any]], so I need to flatmap otherwise the inner IO won't be executed. But IO.flatten would be slightly shorter and consistent with the rest of the code so I'll use IO.flatten instead.

**Comment by wi101** [comment]
@jdegoes Now we can shutdown the queue, 
Do we still need to `interruptTake` and `interruptOffer` ?

**Comment by ghostdogpr** [comment]
@wi101 makes sense, I just added it with a test case

**Comment by jdegoes** [comment]
@wi101 No, they can be private or removed.

File: src/main/scala/scalaz/ioqueue/Queue.scala [hunk 3]
```diff
@@ -152,6 +201,14 @@ class Queue[A] private (capacity: Int, ref: Ref[State[A]]) {
       case d => d
     }.void
 
+  final private def interruptPromise[B](
```
Comment (jdegoes) [inline src/main/scala/scalaz/ioqueue/Queue.scala]: This can be replaced by `p.interrupt0`, at least if we have an `interrupt0` on promise that takes `List[Throwable]`. If not, we should add it.

Comment (ghostdogpr) [reply src/main/scala/scalaz/ioqueue/Queue.scala]: It exists but it's protected. I'll make a PR in ZIO to make it public then.

**Comment by ghostdogpr** [comment]
Removed the 2 unused functions as discussed.

**Comment by ghostdogpr** [comment]
I applied the latest master, solved the conflicts + added the shutdown logic in offerAll.

The last thing to do is replacing `interruptPromise` by `interrupt0` from ZIO. https://github.com/scalaz/scalaz-zio/pull/255 was merged but we need a new release of ZIO to use it. We can wait, or merge and I'll do that change in a later PR.

File: src/test/scala/scalaz/ioqueue/QueueSpec.scala [hunk 4]
```diff
@@ -14,8 +13,6 @@ class QueueSpec(implicit ee: ExecutionEnv) extends AbstractRTSSpec with AroundTi
     Make a Queue and
     add values then call
       `take` to retrieve them in correct order. ${upTo(1.second)(e1)}
-      `interruptTake`to interrupt fiber which is suspended on `take`. ${upTo(1.second)(e2)}
```
Comment (wi101) [inline src/test/scala/scalaz/ioqueue/QueueSpec.scala]: @ghostdogpr the description and implementation of `e2` and `e3` are removed because of the conflict

Comment (ghostdogpr) [reply src/test/scala/scalaz/ioqueue/QueueSpec.scala]: I removed those 2 tests because they were for the now removed `interruptTake` and `interruptOffer`.

**Comment by wi101** [comment]
@ghostdogpr After doing the refactoring , I think we can merge `Shutdown` and after the new release of zio you can do that change in a PR

File: src/test/scala/scalaz/ioqueue/QueueSpec.scala (hunk 4 above)
Comment (ghostdogpr) [reply src/test/scala/scalaz/ioqueue/QueueSpec.scala]: Done!
